File: src/lib/excel/buildMappingRows.ts
import type { MappingRow, SheetData } from '@/types/mapping';
import type { DetectedColumn, MappingFieldKey } from '@/types/columnMapping';

const TRUTHY = new Set(['y', 'yes', 'true', '1', 'x', 'pk']);

function parseBoolean(value: unknown): boolean {
  const text = String(value ?? '').trim().toLowerCase();
  return TRUTHY.has(text);
}

/** Trimmed cell value with any internal newlines/whitespace-runs collapsed to a single space --
 *  safe for any field except sourceField/targetField, which need the raw (pre-split) value. */
function getValue(record: Record<string, unknown>, header: string | null): string {
  if (!header) return '';
  return String(record[header] ?? '')
    .trim()
    .replace(/\s+/g, ' ');
}

/** Trimmed at the ends only -- preserves internal newlines so callers can detect/split multi-value cells. */
function getRawValue(record: Record<string, unknown>, header: string | null): string {
  if (!header) return '';
  return String(record[header] ?? '').trim();
}

/**
 * Splits a cell's raw text on embedded line breaks. Real-world mapping docs sometimes have two
 * field names stacked in one cell via Alt+Enter (a copy/paste or manual-entry artifact) -- if we
 * don't split these apart, the two names get treated as a single garbled field name wherever it's
 * used as a SQL identifier.
 */
function splitMultilineValue(raw: string): string[] {
  return raw
    .split(/\r\n|\r|\n/)
    .map((s) => s.trim())
    .filter(Boolean);
}

/**
 * Some mapping docs list more than one source table in a single Source Table cell (e.g.
 * "srctb1 ,srctb2") when a transformation draws from multiple tables at once (their Source Column
 * cell correspondingly lists "column_1,att_1" -- see businessRuleHeuristics/transformationSql for
 * how those individual column names get recovered for classification). A row's `sourceTable` is
 * used everywhere as a single physical FROM-clause table, though, so only the first is kept here --
 * any additional tables are expected to be documented as joins in the joins sheet instead, which is
 * how their row/schema/referential-integrity checks already get attached.
 */
function firstTableName(value: string): string {
  const [first] = value.split(/[,;]|\band\b|&/i);
  return (first ?? '').trim();
}

export function buildMappingRows(
  sheet: SheetData,
  columns: DetectedColumn<MappingFieldKey>[]
): MappingRow[] {
  const byField = new Map(columns.map((c) => [c.field, c]));
  const get = (field: MappingFieldKey) => byField.get(field) ?? null;

  const pkColumn = get('primaryKeyFlag');
  const nullableColumn = get('nullableFlag');

  return sheet.rows
    .flatMap((row, index) => {
      const sourceFieldRaw = getRawValue(row, get('sourceField')?.matchedHeader ?? null);
      const targetFieldRaw = getRawValue(row, get('targetField')?.matchedHeader ?? null);

      const sourceLines = splitMultilineValue(sourceFieldRaw);
      const targetLines = splitMultilineValue(targetFieldRaw);
      const lineCount = Math.max(sourceLines.length, targetLines.length, 1);

      // A blank cell in a *matched* nullable-flag column is not the same signal as an explicit
      // "N" -- it means this particular row's value was never filled in, not that the analyst
      // affirmatively declared the field NOT NULL. Treating blank the same as an explicit falsy
      // value would fabricate a constraint the document never actually asserted (seen in a real
      // doc where every single row left this column blank, wrongly flagging nearly every field as
      // NOT NULL) -- so a blank cell falls back to the same permissive default as the column being
      // entirely absent, regardless of whether the column is a normal or inverted ("Mandatory") one.
      const nullableCellBlank = !nullableColumn || getValue(row, nullableColumn.matchedHeader ?? null) === '';
      const nullableRaw = nullableColumn ? parseBoolean(row[nullableColumn.matchedHeader ?? '']) : true;
      const isNullable = nullableCellBlank ? true : nullableColumn!.inverted ? !nullableRaw : nullableRaw;

      const shared = {
        sourceTable: firstTableName(getValue(row, get('sourceTable')?.matchedHeader ?? null)),
        sourceSchema: getValue(row, get('sourceSchema')?.matchedHeader ?? null),
        transformation: getValue(row, get('transformation')?.matchedHeader ?? null),
        targetTable: getValue(row, get('targetTable')?.matchedHeader ?? null),
        targetSchema: getValue(row, get('targetSchema')?.matchedHeader ?? null),
        targetDatatype: getValue(row, get('targetDatatype')?.matchedHeader ?? null),
        isPrimaryKey: pkColumn ? parseBoolean(row[pkColumn.matchedHeader ?? '']) : false,
        isNullable,
        sourceFileLocation: getValue(row, get('sourceFileLocation')?.matchedHeader ?? null) || undefined,
        sourceFileName: getValue(row, get('sourceFileName')?.matchedHeader ?? null) || undefined,
        rawRow: row,
        sheetRowNumber: sheet.headerRowIndex + index + 2,
      };

      return Array.from({ length: lineCount }, (_, lineIndex): MappingRow => {
        const sourceField = sourceLines[lineIndex] ?? sourceLines[0] ?? '';
        const targetField = targetLines[lineIndex] ?? targetLines[0] ?? '';
        return {
          id: lineCount > 1 ? `map-${index}-${lineIndex}` : `map-${index}`,
          sourceField,
          targetField,
          ...shared,
        };
      });
    })
    .filter((r) => r.sourceField || r.targetField);
}

File: server/engine/build_mapping_rows.py
"""Direct port of src/lib/excel/buildMappingRows.ts, including the compound-Source-Table and
multi-line-cell fixes from the TypeScript version's regression history.
"""

from __future__ import annotations

import re

from engine.models import DetectedColumn, MappingRow, SheetData

_TRUTHY = {"y", "yes", "true", "1", "x", "pk"}


def _parse_boolean(value: object) -> bool:
    text = str(value if value is not None else "").strip().lower()
    return text in _TRUTHY


def _get_value(record: dict, header: str | None) -> str:
    if not header:
        return ""
    raw = str(record.get(header, "") or "")
    return re.sub(r"\s+", " ", raw.strip())


def _get_raw_value(record: dict, header: str | None) -> str:
    if not header:
        return ""
    return str(record.get(header, "") or "").strip()


def _split_multiline_value(raw: str) -> list[str]:
    return [s.strip() for s in re.split(r"\r\n|\r|\n", raw) if s.strip()]


def _first_table_name(value: str) -> str:
    parts = re.split(r"[,;]|\band\b|&", value, flags=re.IGNORECASE)
    return parts[0].strip() if parts else ""


def build_mapping_rows(sheet: SheetData, columns: list[DetectedColumn]) -> list[MappingRow]:
    by_field = {c.field: c for c in columns}

    def get(field: str) -> DetectedColumn | None:
        return by_field.get(field)

    pk_column = get("primaryKeyFlag")
    nullable_column = get("nullableFlag")

    result: list[MappingRow] = []
    for index, row in enumerate(sheet.rows):
        source_field_raw = _get_raw_value(row, get("sourceField").matched_header if get("sourceField") else None)
        target_field_raw = _get_raw_value(row, get("targetField").matched_header if get("targetField") else None)

        source_lines = _split_multiline_value(source_field_raw)
        target_lines = _split_multiline_value(target_field_raw)
        line_count = max(len(source_lines), len(target_lines), 1)

        nullable_raw = (
            _parse_boolean(row.get(nullable_column.matched_header, "")) if nullable_column and nullable_column.matched_header else True
        )
        is_nullable = (not nullable_raw if nullable_column.inverted else nullable_raw) if nullable_column else True

        shared = {
            "source_table": _first_table_name(_get_value(row, get("sourceTable").matched_header if get("sourceTable") else None)),
            "source_schema": _get_value(row, get("sourceSchema").matched_header if get("sourceSchema") else None),
            "transformation": _get_value(row, get("transformation").matched_header if get("transformation") else None),
            "target_table": _get_value(row, get("targetTable").matched_header if get("targetTable") else None),
            "target_schema": _get_value(row, get("targetSchema").matched_header if get("targetSchema") else None),
            "source_datatype": _get_value(row, get("sourceDatatype").matched_header if get("sourceDatatype") else None),
            "target_datatype": _get_value(row, get("targetDatatype").matched_header if get("targetDatatype") else None),
            "is_primary_key": _parse_boolean(row.get(pk_column.matched_header, "")) if pk_column and pk_column.matched_header else False,
            "is_nullable": is_nullable,
            "source_file_location": _get_value(row, get("sourceFileLocation").matched_header if get("sourceFileLocation") else None) or None,
            "source_file_name": _get_value(row, get("sourceFileName").matched_header if get("sourceFileName") else None) or None,
            "raw_row": row,
            "sheet_row_number": sheet.header_row_index + index + 2,
        }

        for line_index in range(line_count):
            source_field = source_lines[line_index] if line_index < len(source_lines) else (source_lines[0] if source_lines else "")
            target_field = target_lines[line_index] if line_index < len(target_lines) else (target_lines[0] if target_lines else "")
            row_id = f"map-{index}-{line_index}" if line_count > 1 else f"map-{index}"
            mapping_row = MappingRow(id=row_id, source_field=source_field, target_field=target_field, **shared)
            if mapping_row.source_field or mapping_row.target_field:
                result.append(mapping_row)

    return result
